lib/tree.rb

```diff
@@ -63,4 +102,56 @@ def bfs
   def to_s
     return "#{self.inorder}"
   end
+
+  private
+
+  def add_helper(current, new_node)
+    return new_node if current == nil
+
+    if current.key >= new_node.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+
+    return current
+  end
+
+  def inorder_helper(current, tree)
+    return tree if current == nil
+
+    inorder_helper(current.left, tree)
+    tree.push({ :key => current.key, :value => current.value })
+    inorder_helper(current.right, tree)
+
+  end
+
+  def preorder_helper(current, tree)
+    return tree if current == nil
+    
+    tree.push({ :key => current.key, :value => current.value })
+    
+    preorder_helper(current.left, tree)
+    preorder_helper(current.right, tree)
+  end
+
+  def postorder_helper(current, tree)
+    return tree if current == nil
+    
+    postorder_helper(current.left, tree)
+    postorder_helper(current.right, tree)
+
+    tree.push({ :key => current.key, :value => current.value })
+  end
+
+  def height_helper(current, current_height, max_height)
+    return max_height if current == nil
+
+    max_height = current_height if current_height > max_height
+
+    max_height = height_helper(current.left, current_height + 1, max_height)
```

Comment:
I'm not sure you would need to pass in the current_height here, but it does work.